includes/class-newspack-popups-segmentation.php

```diff
@@ -0,0 +1,185 @@
+<?php
+/**
+ * Newspack Segmentation Plugin
+ *
+ * @package Newspack
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Main Newspack Segmentation Plugin Class.
+ */
+final class Newspack_Popups_Segmentation {
+	/**
+	 * The single instance of the class.
+	 *
+	 * @var Newspack_Popups_Segmentation
+	 */
+	protected static $instance = null;
+
+	/**
+	 * Name of the Client ID, to be used by amp-analytics.
+	 */
+	const NEWSPACK_SEGMENTATION_CID_NAME = 'newspack-cid';
+
+	/**
+	 * Query param that will overwrite the cookie value.
+	 */
+	const NEWSPACK_SEGMENTATION_CID_LINKER_PARAM = 'ref_newspack_cid';
+
+	/**
+	 * Main Newspack Segmentation Plugin Instance.
+	 * Ensures only one instance of Newspack Segmentation Plugin Instance is loaded or can be loaded.
+	 *
+	 * @return Newspack Segmentation Plugin Instance - Main instance.
+	 */
+	public static function instance() {
+		if ( is_null( self::$instance ) ) {
+			self::$instance = new self();
+		}
+		return self::$instance;
+	}
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		add_action( 'wp_footer', [ $this, 'insert_amp_analytics' ], 20 );
+		add_action( 'wp_enqueue_scripts', [ __CLASS__, 'wp_enqueue_scripts' ] );
+	}
+
+	/**
+	 * Permission callback for the API calls.
+	 */
+	public static function is_admin_user() {
+		return current_user_can( 'manage_options' );
+	}
+
+	/**
+	 * Should tracking code be inserted?
+	 */
+	public static function is_tracking() {
+		if ( self::is_admin_user() ) {
+			return false;
+		}
+		return true;
+	}
+
+	/**
+	 * Insert amp-analytics scripts.
+	 */
+	public static function wp_enqueue_scripts() {
+		if ( ! self::is_tracking() ) {
+			return;
+		}
+
+		// Register AMP scripts explicitly for non-AMP pages.
+		if ( ! is_admin() && ! wp_script_is( 'amp-runtime', 'registered' ) ) {
+			// phpcs:ignore WordPress.WP.EnqueuedResourceParameters.MissingVersion
+			wp_register_script(
+				'amp-runtime',
+				'https://cdn.ampproject.org/v0.js',
+				null,
+				null,
+				true
+			);
+		}
+		$scripts = [ 'amp-analytics' ];
+		foreach ( $scripts as $script ) {
+			if ( ! wp_script_is( $script, 'registered' ) ) {
+				$path = "https://cdn.ampproject.org/v0/{$script}-latest.js";
+				// phpcs:ignore WordPress.WP.EnqueuedResourceParameters.MissingVersion
+				wp_register_script(
+					$script,
+					$path,
+					array( 'amp-runtime' ),
+					null,
+					true
+				);
+			}
+			wp_enqueue_script( $script );
+		}
+	}
+
+	/**
+	 * Insert amp-analytics tracking code.
+	 * Has to be included on every page to set the cookie.
+	 */
+	public function insert_amp_analytics() {
```

Comment:
Sorry for being thick, but I'm still a bit unclear on why we need to make this `amp-analytics` `POST`, as opposed to using the `amp-access` `GET` to write the `post_read` data to the the DB (or log file after https://github.com/Automattic/newspack-popups/pull/233 is merged in)? Reducing these down to one would obviously be a big performance win.

Reply:
While trying to reduce it to one request I've found our that we have a problem here anyway. It gets complicated so here's a tl;dr:

> When navigating from proxied site (AMP Cache) to origin site, the **first request** to `amp-access`' authorisation endpoint will receive a **wrong client ID**. Subsequent requests will be ok.

In practice, this means that some segmentation features will not kick in on this single request, because `amp-access` will be querying wrong data (no data, to be exact).

It boils down to:

## Cross-context linking and interoperation between `amp-access` and `amp-analytics`. 

Considering the following scenario:
1. User visits the page served from AMP Cache (proxied site)
2. User clicks a link and thus navigates to the origin site

For `amp-access` these will be visits from **two separate clients**, because a proxied site's cookie cannot be read by the origin site. Now, this same problem is resolved by `amp-analytics` with [`linker` configuration](https://amp.dev/documentation/components/amp-analytics/?format=websites#linkers). With it, any link on the proxied site gets a special param appended, with an encoded client ID. On loading the origin site, `amp-analytics` reads this param and sets it in a cookie. 

The issue with interoperability between these two AMP components I've run into manifests itself like this:

1. Page is loaded from AMP Cache. Client is assigned a client id, let's call it `id-A`. This is encoded and appended to URLs leading to origin site.
2. User navigates to the origin site. 
3. `amp-access`, seeing no client ID in site's cookies, assigns a client ID `id-B` and uses it to call the authorisation endpoint
4. `amp-analytics` reads the linker URL param, sets the `id-A` in cookie
5. Subsequent calls to `amp-access`' endpoint are made with `id-A` client ID (correctly), because it's read from the cookie

I _guess_ this issue could be resolved by either: 
1. Reversing the order of events – 4. first, 3. then. However I was not able to do that, because `amp-analytics` is placed in the document body, while `amp-access` in document head
2. Having that `linker` feature in `amp-access`

## What would change with a single request

This first page load after navigating from AMP Cache – now serving in place of analytics – will assign this visit to a wrong client ID. This means that this page load will be "lost", and because this client ID will not be used anymore, create a superfluous row in the DB. 

---

I'll research it a bit more, but if my conclusions here are correct, we have a situation where we have to choose an extra 
API request or extra rows in the DB. From a reader's perspective, the way it is now is better, because the data reported will be correct, even though they might get a different experience on that first-origin-page-load-after-amp-cache then intended by the publisher.

Reply:
Added visit-reporting handling to the amp-access endpoint in e2b1f30de7841e1063c77fb18c578cde1b964482